specs/Print.md

```diff
@@ -0,0 +1,816 @@
+
+Developers have requested the ability to programmatically print the current page optionally
+using a print dialog in WebView2. Currently the end user can open a print preview dialog by
+pressing "Ctrl+ P" or selecting "Print" from the context menu in a document or can open the
+system print dialog by pressing "Ctrl + Shift + P" or clicking the "Print using system dialog"
+link in the print preview dialog. This API gives developers the ability to open the print dialog
+without requiring the end user interaction and also can print without a dialog to a specific
+printer by configuring the print settings.
+
+You can use `CoreWebView2.PrintToPdfAsync` API to print a current page as a pdf file to disk and
+then use native printing APIs to print the document to a specific printer. However, this API
+requires the PDF to be printed to the disk which can be prohibitive at times for some financia
+customers if the printed information is sensitive and also printing to printer without a dialog
+requires additional effort on the developers for cross platform apps.
+
+In this document we describe the updated API. We'd appreciate your feedback.
+
+# Description
+We propose following APIs
+
+**ShowPrintUI API**: This API gives developers the ability to open a browser print preview dialog or
+system print dialog programmatically. This API consists of a ShowPrintUI method and can either open
+a browser print preview dialog or system print dialog based on COREWEBVIEW2_PRINT_DIALOG_KIND value.
+
+**Print**: This API gives developers the ability to print the current web page to the specific printer
+without a print dialog. This API consists of an asynchronous Print method and a PrintSettings2 object.
+PrintSettings2 object extends PrintSettings to support printing to a specific printer.
+
+**Print to PDF Stream**: This API gives developers printable PDF data, which can be used as a preview
+in a custom print preview dialog or print pdf data of current web page using native printer APIs.
+This API consists of an asynchronous PrintToPdfStream method and a PrintSettings2 object.
+
+
+# Examples
+## ShowPrintUI
+
+You can use the `ShowPrintUI` method to programmatically open a print dialog. Currently devs are
+using JavaScript `window.print()` using the `CoreWebView2.ExecuteScriptAsync` API to invoke the
+browser print preview dialog. But there isn't a way to invoke the system print dialog
+programmatically nor will `window.print()` work if script is disabled on the page, or the
+document is a PDF or other non-HTML document.
+
+```cpp
+// This example shows the user a print dialog. If `printDialogKind` is
+// COREWEBVIEW2_PRINT_DIALOG_KIND_BROWSER, opens a browser print preview dialog,
+// otherwise opens a system print dialog.
+
+bool AppWindow::ShowPrintUI(COREWEBVIEW2_PRINT_DIALOG_KIND printDialogKind)
+{
+  wil::com_ptr<ICoreWebView2_15> webView2_15;
+  CHECK_FAILURE(m_webView->QueryInterface(IID_PPV_ARGS(&webView2_15)));
+  CHECK_FEATURE_RETURN(webView2_15);
```

Comment:
This line has no effect because `CHECK_FAILURE` is going to throw if the QI fails. Presumably, we want a `try_query<>` or some other non-throwing mechanism.

```cpp
wil::com_ptr<ICoreWebView2_15> webView2_15 = m_webView.try_query<ICoreWebView2_15>();
CHECK_FEATURE_RETURN(webView2_15);
```

Reply:
In [previous comment] (https://github.com/MicrosoftEdge/WebView2Feedback/pull/2604#discussion_r930372556), we decided to do CHECK_FAILURE for this and remove CHECK_FEATURE_RETURN part. Looks like I haven't updated this one.